core/arch/arm/kernel/spmc_sp_handler.c

```diff
@@ -0,0 +1,220 @@
+{
+	if (args->a2 != FFA_PARAM_MBZ) {
+		/* a2 should always be 0 */
+		*dst = NULL;
+		return FFA_INVALID_PARAMETERS;
+	}
+
+	*dst = sp_get_session(FFA_DST(args->a1));
+
+	/* Message came from the NW */
+	if (!caller) {
+		if (!*dst) {
+			EMSG("Neither destination nor source is a SP!");
+			return FFA_INVALID_PARAMETERS;
+		}
+	} else {
+		/* Check if the source matches the endpoint we came from */
+		if (FFA_SRC(args->a1) != caller->endpoint_id) {
+			EMSG("Source address doesn't match the endpoint id!");
+			return FFA_INVALID_PARAMETERS;
+		}
+	}
+	return FFA_OK;
+}
+
+static struct sp_session *
+ffa_handle_sp_direct_req(struct thread_smc_args *args,
+			 struct sp_session *caller_sp)
+{
+	struct sp_session *dst = NULL;
+	TEE_Result res = FFA_OK;
+
+	if (args->a2 != FFA_PARAM_MBZ) {
+		/* a2 should always be 0 */
+		dst = NULL;
+		ffa_set_error(args, FFA_INVALID_PARAMETERS);
+		return NULL;
+	}
+
+	res = ffa_get_dst(args, caller_sp, &dst);
+	if (res) {
+		/* Tried to send message to an incorrect endpoint */
+		ffa_set_error(args, res);
+		return caller_sp;
+	}
+	if (!dst) {
+		EMSG("Request to normal world not supported!");
+		ffa_set_error(args, FFA_NOT_SUPPORTED);
+		return NULL;
+	}
+
+	cpu_spin_lock(&dst->spinlock);
+	if (dst->state != sp_idle) {
+		EMSG("SP is busy");
+		ffa_set_error(args, FFA_BUSY);
+		cpu_spin_unlock(&dst->spinlock);
+		return caller_sp;
+	}
+
+	dst->state = sp_busy;
+	cpu_spin_unlock(&dst->spinlock);
+
+	/*
+	 * Store the calling endpoint id. This will make it possible to check
+	 * if the response is send back to the correct endpoint.
+	 */
+	dst->caller_id = FFA_SRC(args->a1);
+
+	/* Forward the message to the destination Secure Partition. */
+	res = sp_enter(args, dst);
+	if (res) {
+		/* The SP Panicked */
+		ffa_set_error(args, FFA_ABORTED);
+		/* Return error to calling SP */
+		return caller_sp;
+	}
+	return dst;
+}
+
+static struct sp_session *
+ffa_handle_sp_direct_resp(struct thread_smc_args *args,
+			  struct sp_session *caller_sp)
+{
+	struct sp_session *dst = NULL;
+	TEE_Result res = FFA_OK;
+
+	if (!caller_sp) {
+		EMSG("Response from normal world not supported!");
+		ffa_set_error(args, FFA_NOT_SUPPORTED);
+		return NULL;
+	}
+
+	res = ffa_get_dst(args, caller_sp, &dst);
+	if (res) {
+		/* Tried to send response to an incorrect endpoint */
+		ffa_set_error(args, res);
+		return caller_sp;
+	}
+
+	if (caller_sp->state != sp_busy) {
+		EMSG("SP is not waiting for a request!");
+		ffa_set_error(args, FFA_INVALID_PARAMETERS);
+		return caller_sp;
+	}
+
+	if (caller_sp->caller_id != FFA_DST(args->a1)) {
+		EMSG("FFA_MSG_SEND_DIRECT_RESP to incorrect SP!");
+		ffa_set_error(args, FFA_INVALID_PARAMETERS);
+		return caller_sp;
+	}
+
+	caller_sp->caller_id = 0;
+
+	cpu_spin_lock(&caller_sp->spinlock);
+	caller_sp->state = sp_idle;
+	cpu_spin_unlock(&caller_sp->spinlock);
+
+	if (!dst) {
+		/* Send message back to the NW */
+		return dst;
+	}
+
+	/* Forward the message to the destination Secure Partition */
```

Comment:
s/Secure Partition/SP/
for consistency.

Reply:
Done